R/util_dfc.R

```diff
@@ -0,0 +1,238 @@
+###     Dan Greene, 12/21/17
+###
+###     The goal of the util_dfc ("data frame compare") module is to make it easy to see
+###     where two data frames differ for debugging and further analysis.
+###
+###     Depends on util.R
+###
+
+util_dfc.compare_vecs <- function(vec1, vec2) {
+  # compare vector elements without regard to order and without checking for duplicates.
+  # return a list with three components -
+  ## shared_elements: the shared elements,
+  ## only_in_first: the elements in vec1 and not vec2,
+  ## only_in_second: the elements in vec2 and not vec1.
+  return(list(shared_elements = intersect(vec1, vec2),
+              only_in_first = setdiff(vec1, vec2),
+              only_in_second = setdiff(vec2, vec1)))
+}
+
+
+util_dfc.get_condensed_ids <- function(df, id_cols) {
+  # Helper function for getting a vector of IDs (not necessarily unique) from a DF.
+  # If there are multiple ID columns, they need to be concatenated with "__".
+  if(length(id_cols) == 1) {
+    return(as.character(df[, id_cols]))
+  } else {
+    return(apply(df[, id_cols], 1, paste, collapse = "__"))
+  }
+}
+
+
+util_dfc.compare_colnames <- function(df1, df2) {
+  # compare the column names of two data frames using util_dfc.compare_vecs.
+  # return a list with three components -
+  ## shared_elements: the shared names,
+  ## only_in_first: the names in df1 and not df2,
+  ## only_in_second: the names in df2 and not df1
+
+  # Sanity checks
+  if(!is.data.frame(df1) | !is.data.frame(df2)) {
+    stop("Error - at least one of the first two arguments for util_dfc.compare_colnames is not a data frame.")
+  }
+
+  return(util_dfc.compare_vecs(names(df1), names(df2)))
+}
+
+
+
+
+util_dfc.compare_identifiers <- function(df1, df2, id_cols, id_cols_uniquely_identify_rows = FALSE) {
+  # Compare two dfs on a set of one or more commonly-shared columns that identify rows.
+  # If rows are identified by multiple columns (e.g. team x student x week), include all column names in id_cols.
+  # Return a list with shared IDs, IDs in df1 only, and IDs in df2 only.
+  #
+  # If you expect id_cols to UNIQUELY identify each row in both dfs, set id_cols_uniquely_identify_rows to TRUE.
+  # Otherwise, duplicate ID values within each data frame will get ignored.
+  # Example use case: you want to compare a student-x-week df with a student df to see if they have the same students.
+  # The commonly-shared id_col is student ID.
+  # Since student ID doesn't uniquely identify rows in the first df, id_cols_uniquely_identify_rows = FALSE.
+
+  # Sanity checks
+  if(!is.data.frame(df1) | !is.data.frame(df2)) {
+    stop("Error - at least one of the first two arguments for util_dfc.compare_unique_identifiers is not a data frame.")
+  }
+  if(is.na(id_cols) | length(id_cols) == 0) {
+    stop("Error - id_cols argument is NA or empty.")
+  }
+  if(any(!id_cols %in% names(df1)) | any(!id_cols %in% names(df2))) {
+    stop("Error - not all id columns provided are in both data frames.")
```

Comment:
I always make the error message say which values are missing, e.g.,

```
missing_cols <- id_cols[!id_cols %in% names(df)]
stop("Error - The following id columns provided were missing from one or more data.frames:" %+%
    paste0(missing_cols, collapse = ", ")
```

Reply:
Yes!